lib/resumption/client.go

```diff
@@ -0,0 +1,362 @@
+	maxBackoff = 10 * time.Second
+)
+
+var resumablePreludeLine = regexp.MustCompile(`^` +
+	regexp.QuoteMeta(serverProtocolStringV1) +
+	` ([0-9A-Za-z+/]{` + strconv.Itoa(base64.RawStdEncoding.EncodedLen(ecdhP256UncompressedSize)) + `}) ` + `([0-9a-z\-]+)\r\n$`)
+
+// readServerVersionExchange returns the ECDH public key and the host ID
+// extracted from a resumption v1 server version line. A triplet of (nil, "",
+// nil) is returned if a server version line is peeked and is not a resumption
+// v1 line.
+func readServerVersionExchange(conn *multiplexer.Conn) (dhPubKey *ecdh.PublicKey, hostID string, err error) {
+	const maxVersionIdentifierSize = 255
+	line, err := peekLine(conn, maxVersionIdentifierSize)
+	if err != nil {
+		return
+	}
+
+	match := resumablePreludeLine.FindSubmatch(line)
+	if match == nil {
+		return nil, "", nil
+	}
+
+	var buf [ecdhP256UncompressedSize]byte
+	if n, err := base64.RawStdEncoding.Decode(buf[:], match[1]); err != nil {
+		return nil, "", trace.Wrap(err)
+	} else if n != ecdhP256UncompressedSize {
+		return nil, "", trace.Wrap(io.ErrUnexpectedEOF, "short ECDH encoding")
+	}
+
+	dhPubKey, err = ecdh.P256().NewPublicKey(buf[:])
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	hostID = string(match[2])
+
+	// discard is guaranteed to work for the line we just peeked
+	_, _ = conn.Discard(len(line))
+
+	return
+}
+
+// redialFunc should dial the given host; the connection is allowed to die with
+// the passed context (to accommodate the Teleport gRPC transport).
+type redialFunc = func(ctx context.Context, hostID string) (net.Conn, error)
+
+// WrapSSHClientConn tries to detect if the server at the other end of nc is a
+// resumption v1 server, and if so it returns a [net.Conn] that will
+// transparently resume itself (using the provided redial func). If the
+// connection is wrapped, the context applies to the lifetime of the returned
+// connection, not just the duration of the function call.
+func WrapSSHClientConn(ctx context.Context, nc net.Conn, redial redialFunc) (net.Conn, error) {
+	dhKey, err := ecdh.P256().GenerateKey(rand.Reader)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to generate ECDH key, proceeding without resumption (this is a bug).")
+		return nc, nil
+	}
+
+	// adds a read buffer with the ability to peek to nc
+	conn := ensureMultiplexerConn(nc)
+
+	// We must send the first 8 bytes of the version string to go through some
+	// older versions of the multiplexer that sits in front of the Teleport SSH
+	// server; thankfully, no matter which SSH client we'll end up using, it
+	// must send `SSH-2.0-` as its first 8 bytes, as per RFC 4253 ("The Secure
+	// Shell (SSH) Transport Layer Protocol") section 4.2. Sending only the
+	// first 8 bytes rather than a full version string is noncompliant behavior,
+	// but our SSH client is only intended to talk to Teleport-implemented SSH
+	// servers anyway, and other clients in the ecosystem do much worse
```

Comment:
Unless I misunderstand how this code is being used, won't this also be hit when dialing openssh agents?

Reply:
Modern `tsh` never actually ends up speaking to an OpenSSH `sshd`, it's always either the SSH server on a Teleport SSH agent or the forwarding SSH server on a Teleport Proxy.